12/umbraco-cms/reference/routing/custom-middleware.md

```diff
@@ -0,0 +1,70 @@
+---
+description: "Customizing the ASP.NET middleware pipeline in Umbraco"
+---
+
+# Middleware
+
+Middleware is a way of handling requests and responses before or after they are completed. An example can be checking if a user is logged in before adding an item to the cart. In this example it checks something before an http request is made.
+
+Umbraco automatically configures all required middleware in the `WithMiddleware()` method in a specific order based on the [Microsoft documentation](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-7.0#middleware-order).
+
+You can use Umbraco pipeline filters in case you want to add your own middleware before, in-between or after the default Umbraco middleware. Filters are added by configuring the `UmbracoPipelineOptions` and require an instance of `IUmbracoPipelineFilter` that contains the following callbacks:
+
+- `PrePipeline` - executed before any Umbraco-specific middleware is added, an example can be [URL rewrites](iisrewriterules.md).
+- `PreRouting` - executed after the static files middleware and before the [routing middleware](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-7.0) is added (using `UseRouting()`). It can also be used to change the incoming URL.
+- `PostRouting` - executed after the routing middleware is added and can be used to [configure CORS](https://learn.microsoft.com/en-us/aspnet/core/security/cors?view=aspnetcore-7.0).
+- `PostPipeline` - executed after all Umbraco-specific middleware is added.
+- `Endpoints` - executed just before the Umbraco-specific endpoints are added using `WithEndpoints()`.
+
+The addition of the `PostRouting` callback is to allow correctly configuring the Cross-Origin Resource Sharing (CORS) middleware without having to use the `WithCustomMiddleware()` method.
+
+- `IUmbracoPipelineFilter` is an interface in Umbraco that allows the creation of custom filters which then modifies the behavior of the request pipeline. It can be used to change different aspects of how Umbraco handles incoming requests, such as changing content or adding security checks.
+- `WithCustomMiddleware()` is a method that can be used in Umbraco for adding custom middleware. This includes some specific customizable instructions that run in the request processing pipeline. This method is often used in combination with IUmbracoPipelineFilter.
+
+{% hint style="warning" %}
+
+Using `WithCustomMiddleware()` instead of `WithMiddelware()` should only be used as a last resort, as Umbraco can break if you forget to add middleware or add them in the wrong order.
```

Comment:
```suggestion
Using `WithCustomMiddleware()` instead of `WithMiddelware()` should only be used as a last resort. This is because Umbraco can break if you forget to add middleware or add them in the wrong order.
```